### PR TITLE
fix(gh-design): skip self-review requests and tolerate 422 errors

### PR DESCRIPTION
**Pull request by snomiao**
## Summary

The 5-minute [`Combined GitHub Tasks`](https://github.com/Comfy-Org/Comfy-PR/actions/runs/24778612522) workflow has been failing every run. Root cause:

```
GitHub Design Task: 21128ms - Review cannot be requested from pull request author.
1 task(s) failed. Exiting with error code 1.
```

`REQUEST_REVIEWERS` is hardcoded to `["PabloWiedemann"]` in `app/tasks/gh-design/gh-design.ts`. When Pablo opens a PR himself, the task calls `gh.pulls.requestReviewers` with him as the reviewer, GitHub responds **422 Review cannot be requested from pull request author**, and `run-gh-tasks.ts` then exits non-zero — turning one avoidable PR-level error into a full workflow failure on every 5-minute tick.

## Fix

- **Skip self-review**: filter the PR author (`task.user`) out of the reviewer list before calling the API (case-insensitive comparison).
- **Tolerate 422**: wrap the call in a try/catch that swallows only `status === 422` (non-collaborators, etc.), rethrowing anything else. 422s are retried on subsequent runs since the condition may be transient.
- **Persist only on success**: reviewer list is saved to DB only after a successful API call, keeping DB state accurate.
- **No redundant retries**: the guard condition uses the filtered `newReviewers` list, so when all eligible reviewers are already requested, the block is skipped entirely.
- **Extracted `filterReviewers()`** into a side-effect-free module with 8 unit tests covering author filtering, case insensitivity, and edge cases.

## Test plan

- [x] CI `run_comfy_pr` and `run_combined_github_tasks` pass
- [x] `filterReviewers.spec.ts` — 8 tests passing
- [ ] Next scheduled `Combined GitHub Tasks` run finishes green.
- [ ] PRs by non-authors still get reviewer requests as before.
- [ ] PRs authored by a `REQUEST_REVIEWERS` member skip the call silently.

### CODE REVIEW

File: app/tasks/gh-design/gh-design.ts [hunk 1]
```diff
@@ -259,15 +259,27 @@ export async function runGithubDesignTask() {
...
+              try {
+                await gh.pulls.requestReviewers({
+                  owner,
+                  repo,
+                  pull_number: issue_number,
+                  reviewers: newReviewers,
+                });
+              } catch (err: any) {
+                // GitHub returns 422 when the requested reviewer is the PR author
+                // or otherwise cannot be added. Record the attempt to avoid
```
[the remark this answers is not in the document]
Comment (snomiao) [reply app/tasks/gh-design/gh-design.ts]: Applied in f97f018 — the comment now accurately describes the remaining 422 cases (non-collaborator, cannot be requested, already requested) since self-review is filtered out beforehand.

File: app/tasks/gh-design/gh-design.ts [hunk 2]
```diff
@@ -259,15 +259,27 @@ export async function runGithubDesignTask() {
...
+                // retrying on every 5-minute schedule run.
+                if (err?.status !== 422) throw err;
+                tlog(`Reviewer request rejected (422): ${err.message}`);
+              }
```
[the remark this answers is not in the document]
Comment (snomiao) [reply app/tasks/gh-design/gh-design.ts]: Added in f97f018 — extracted filterReviewers() as a pure function and added filterReviewers.spec.ts with 6 tests covering: self-authored PRs (empty list), already-requested reviewers, mixed scenarios, and undefined alreadyRequested.

File: app/tasks/gh-design/gh-design.ts [hunk 3]
```diff
@@ -259,15 +259,27 @@ export async function runGithubDesignTask() {
           REQUEST_REVIEWERS.some((e) => !task.reviewers?.includes(e))
         ) {
           const requestReviewers = REQUEST_REVIEWERS;
-          const newReviewers = requestReviewers.filter((e) => !task.reviewers?.includes(e));
-          tlog(`Requesting reviewers: ${newReviewers.join(", ")}`);
+          const newReviewers = requestReviewers.filter(
+            (e) => !task.reviewers?.includes(e) && e !== task.user,
+          );
+          tlog(`Requesting reviewers: ${newReviewers.join(", ") || "(none)"}`);
```
[the remark this answers is not in the document]
Comment (snomiao) [reply app/tasks/gh-design/gh-design.ts]: Fixed in f97f018 — requestReviewers now filters out task.user before use, so persisted data accurately reflects only the reviewers who were actually eligible to be requested.

File: app/tasks/gh-design/gh-design.ts [hunk 4]
```diff
@@ -259,15 +259,27 @@ export async function runGithubDesignTask() {
...
+                  repo,
+                  pull_number: issue_number,
+                  reviewers: newReviewers,
+                });
+              } catch (err: any) {
+                // GitHub returns 422 when the requested reviewer is the PR author
+                // or otherwise cannot be added. Record the attempt to avoid
+                // retrying on every 5-minute schedule run.
+                if (err?.status !== 422) throw err;
+                tlog(`Reviewer request rejected (422): ${err.message}`);
```
[the remark this answers is not in the document]
Comment (snomiao) [reply app/tasks/gh-design/gh-design.ts]: Fixed in f97f018 — changed catch (err: any) to catch (err: unknown) with structural narrowing via (err as { status?: number })?.status.

File: app/tasks/gh-design/filterReviewers.spec.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
+import { describe, expect, it } from "bun:test";
+import { filterReviewers } from "./gh-design";
+
+describe("filterReviewers", () => {
+  const REVIEWERS = ["PabloWiedemann", "AliceDev"];
+
```
[the remark this answers is not in the document]
Comment (snomiao) [reply app/tasks/gh-design/filterReviewers.spec.ts:6]: Fixed in cee9996 — extracted filterReviewers() into its own side-effect-free module (filterReviewers.ts). Test now imports directly from it without triggering any DB initialization. Test time dropped from ~1126ms to ~148ms.

File: app/tasks/gh-design/gh-design.ts [hunk 6]
```diff
@@ -258,16 +272,31 @@ export async function runGithubDesignTask() {
...
           REQUEST_REVIEWERS.some((e) => !task.reviewers?.includes(e))
         ) {
-          const requestReviewers = REQUEST_REVIEWERS;
-          const newReviewers = requestReviewers.filter((e) => !task.reviewers?.includes(e));
-          tlog(`Requesting reviewers: ${newReviewers.join(", ")}`);
+          const { requestReviewers, newReviewers } = filterReviewers(
+            REQUEST_REVIEWERS,
+            task.user,
+            task.reviewers,
+          );
```
[the remark this answers is not in the document]
Comment (snomiao) [reply app/tasks/gh-design/gh-design.ts]: Fixed in cee9996 — removed the REQUEST_REVIEWERS.some(...) guard and replaced it with newReviewers.length > 0 check after computing the filtered list. Now when the PR author is in REQUEST_REVIEWERS and has already been filtered out, the block won't re-enter and re-save on every tick.

File: app/tasks/gh-design/gh-design.ts [hunk 7]
```diff
@@ -254,21 +255,33 @@ export async function runGithubDesignTask() {
...
+              } catch (err: unknown) {
+                // GitHub may return 422 when a requested reviewer cannot be added,
+                // such as when they are not a collaborator, cannot be requested,
+                // or have already been requested. Record the attempt to avoid
+                // retrying on every 5-minute schedule run.
+                const status = (err as { status?: number })?.status;
+                if (status !== 422) throw err;
+                tlog(`Reviewer request rejected (422): ${err}`);
+              }
+              task = await saveGithubDesignTask(url, { reviewers: requestReviewers });
```
[the remark this answers is not in the document]
Comment (snomiao) [reply app/tasks/gh-design/gh-design.ts]: Fixed — reviewers are now only persisted to DB on successful API call. On 422 rejection, the task will retry on the next run.

File: app/tasks/gh-design/filterReviewers.ts [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+ * Filter the reviewer list for a PR: excludes the PR author and
+ * anyone who has already been requested.
+ */
+export function filterReviewers(
+  allReviewers: string[],
+  prAuthor: string,
+  alreadyRequested?: string[],
+): { requestReviewers: string[]; newReviewers: string[] } {
+  const requestReviewers = allReviewers.filter((e) => e !== prAuthor);
+  const newReviewers = requestReviewers.filter((e) => !alreadyRequested?.includes(e));
```
[the remark this answers is not in the document]
Comment (snomiao) [reply app/tasks/gh-design/filterReviewers.ts]: Fixed in e43396f — filterReviewers() now normalizes usernames to lowercase for all comparisons while preserving original casing in the returned arrays. Added 2 tests for case-insensitive matching.

File: app/tasks/gh-design/gh-design.ts [hunk 9]
```diff
@@ -254,21 +255,37 @@ export async function runGithubDesignTask() {
...
+                reviewersRequested = true;
+              } catch (err: unknown) {
+                // GitHub may return 422 when a requested reviewer cannot be added,
+                // such as when they are not a collaborator, cannot be requested,
+                // or have already been requested. Record the attempt to avoid
+                // retrying on every 5-minute schedule run.
+                const status = (err as { status?: number })?.status;
+                if (status !== 422) throw err;
+                tlog(`Reviewer request rejected (422): ${err}`);
+              }
```
[the remark this answers is not in the document]
Comment (snomiao) [reply app/tasks/gh-design/gh-design.ts:284]: Fixed in 29caf1f — updated the comment to accurately state that 422 rejections are retried on subsequent runs (not suppressed), since the reviewer may become eligible later. This is intentional: the primary self-review case is already filtered out, and remaining 422s (e.g., non-collaborator) are transient conditions worth retrying.

File: app/tasks/gh-design/filterReviewers.ts [hunk 10]
```diff
@@ -0,0 +1,16 @@
+/**
+ * Filter the reviewer list for a PR: excludes the PR author and
+ * anyone who has already been requested.
```
[the remark this answers is not in the document]
Comment (snomiao) [reply app/tasks/gh-design/filterReviewers.ts]: Fixed in add499b — updated the docstring to accurately describe the two return values: requestReviewers (minus author) and newReviewers (minus author and already-requested).

File: app/tasks/gh-design/gh-design.ts [hunk 11]
```diff
@@ -254,21 +255,37 @@ export async function runGithubDesignTask() {
...
+                // such as when they are not a collaborator or cannot be requested.
+                // We log but don't persist, so the request will be retried on the
+                // next run (the reviewer may become eligible later).
+                const status = (err as { status?: number })?.status;
+                if (status !== 422) throw err;
+                tlog(`Reviewer request rejected (422): ${err}`);
+              }
+              if (reviewersRequested) {
+                task = await saveGithubDesignTask(url, { reviewers: requestReviewers });
+              }
```
[the remark this answers is not in the document]
Comment (snomiao) [reply app/tasks/gh-design/gh-design.ts:287]: Fixed — updated PR description to accurately describe the persistence behavior: reviewers are only saved on successful API call, and 422s are retried on subsequent runs.
